omnigibson/transition_rules.py

```diff
@@ -20,37 +40,205 @@
+
+        Args:
+            rules (list of BaseTransitionRule): List of transition rules whose candidate lists should be refreshed
+            objects (list of BaseObject): List of objects that will be used to infer which transition rules are
+                currently active
+        """
+        for rule in rules:
+            # Check if rule is still valid, if so, update its entry
+            candidates = cls.get_rule_candidates(rule=rule, objects=objects)
+
+            # Update candidates if valid, otherwise pop the entry if it exists in cls.ACTIVE_RULES
+            if candidates is not None:
+                # We have a valid rule which should be active, so grab and initialize all of its conditions
+                rule.initialize()
+                conditions = []
+                combined_candidates = dict(**candidates["individual"], **candidates["group"])
+                for condition in rule.conditions:
+                    condition.initialize(combined_candidates)
+                    conditions.append(condition)
+                cls.ACTIVE_RULES[rule] = {
+                    "conditions": conditions,
+                    "candidates": candidates,
+                }
+            elif rule in cls.ACTIVE_RULES:
+                cls.ACTIVE_RULES.pop(rule)
+
+    @classmethod
+    def step(cls):
+        """
+        Steps all active transition rules, checking if any are satisfied, and if so, executing their transitions
+        """
+        # First apply any transition object init states from before, and then clear the dictionary
+        for obj, states_info in cls._INIT_STATES.items():
+            for state, args in states_info.items():
+                obj.states[state].set_value(*args)
+        cls._INIT_STATES = dict()
+
+        # Iterate over all active rules and process the rule for every valid object candidate combination
+        # Cast to list before iterating since ACTIVE_RULES may get updated mid-iteration
+        added_obj_attrs = []
+        removed_objs = []
+        for rule in tuple(cls.ACTIVE_RULES.keys()):
+            # Compile candidates (shallow) copy that will be mutated in place by conditions
+            conditions, candidates = cls.ACTIVE_RULES[rule]["conditions"], cls.ACTIVE_RULES[rule]["candidates"]
+            combined_candidates = dict(**candidates["individual"], **candidates["group"])
+            should_transition = True
+            for condition in conditions:
+                if not condition(object_candidates=combined_candidates):
+                    # Condition was not met, so break early
+                    should_transition = False
+                    break
+            if should_transition:
+                # Compile all valid transitions, and take the transition
+                pruned_individual_candidates = {k: combined_candidates[k] for k in candidates["individual"].keys()}
+                group_objects = {k: combined_candidates[k] for k in candidates["group"].keys()}
```

Comment:
Sorry it's a bit hard to understand what these lines are doing.

Reply:
- only have "filters", and have separate property individual_filter_names
- move ACTIVE_RULE information into the rule directly since it's monolithic